scipy/optimize/_numdiff.py

```diff
@@ -289,8 +377,62 @@ def fun_wrapped(x):
     return J_transposed.T
 
 
+def _sparse_difference(fun, x0, f0, h, use_one_sided,
+                       structure, groups, method):
+    m = f0.size
```

Comment:
Some words of explanation of what's going on here and in group_columns would be helpful.